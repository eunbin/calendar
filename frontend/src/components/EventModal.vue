<template>
  <div class="modal"
       :class="{ show: value }">
    <div class="modal__inner">
      <button class="modal__close" @click="close"></button>
      <h2>일정 {{ isAdd ? '등록' : '수정'}}</h2>
      <div class="form">
        <div>
          <label for="title">일정 제목</label>
          <input v-model="model.title" id="title" type="text">
        </div>
        <div>
          <label for="startDate">시작시간</label>
          <date-pick v-model="model.start.dateTime"
                     start-week-on-sunday
                     pick-time
                     :format="datePickerFormat"
                     id="startDate"
                     @input="onChangeStartDateTime"></date-pick>
        </div>
        <div>
          <label for="endDate">종료시간</label>
          <date-pick v-model="model.end.dateTime"
                     start-week-on-sunday
                     pick-time
                     :format="datePickerFormat"
                     id="endDate"
                     @input="onChangeEndDateTime"></date-pick>
        </div><div>
      </div>
      </div>
      <p v-if="errors.length"
         class="validation">
        <b>아래 에러를 확인해주세요.</b>
        <ul>
          <li v-for="(error, index) in errors"
              :key="index">{{ error }}
          </li>
        </ul>
      </p>
      <div class="modal__action">
        <button @click="close">취소</button>
        <button v-show="!isAdd" class="accent" @click="deleteEvent">삭제</button>
        <button class="info" @click="submit">{{ isAdd ? '등록' : '저장'}}</button>
      </div>
    </div>
  </div>
</template>

<script>
import DatePick from 'vue-date-pick'
import 'vue-date-pick/dist/vueDatePick.css'
import moment from 'moment'
import { cloneDeep } from 'lodash-es'
// import { dateFormat } from '@/date'

export default {
  name: 'EventModal',
  components: { DatePick },
  props: {
    value: {
      type: Boolean,
      default: false
    },
    event: {
      type: Object
    }
  },
  data () {
    return {
      model: null,
      errors: []
    }
  },
  computed: {
    isAdd () {
      return !this.event.id
    },
    datePickerFormat () {
      return 'YYYY-MM-DD HH:mm'
    }
  },
  watch: {
    event: {
      immediate: true,
      handler () {
        this.model = cloneDeep(this.event)
        this.errors = []
      }
    }
  },
  created () {
    this.addEventListeners()
  },
  beforeDestroy () {
    this.removeEventListeners()
  },
  methods: {
    onChangeStartDateTime (val) {
      const startDateTime = moment(val)
      const endDateTime = startDateTime.clone().add(1, 'hour')
      this.model.start.date = startDateTime.format('YYYY-MM-DD')
      this.model.end.date = endDateTime.format('YYYY-MM-DD')
      this.model.end.dateTime = endDateTime.format('YYYY-MM-DD HH:mm')
    },
    onChangeEndDateTime (val) {
      const endDateTime = moment(val)
      const startDateTime = endDateTime.clone().subtract(1, 'hour')
      this.model.end.date = endDateTime.format('YYYY-MM-DD')
      this.model.start.date = startDateTime.format('YYYY-MM-DD')
      this.model.start.dateTime = startDateTime.format('YYYY-MM-DD HH:mm')
    },
    submit () {
      if (this.isValidForm()) {
        this.$emit('submit', this.model, this.isAdd)
      }
    },
    deleteEvent () {
      this.$emit('delete')
    },
    close () {
      this.$emit('close')
    },
    isValidForm () {
      if (this.model.title && this.model.start && this.model.end) {
        return true
      }
      this.errors = []
      if (!this.model.title) {
        this.errors.push('일정 제목 입력은 필수입니다.')
      }
      if (!this.model.start) {
        this.errors.push('시작날짜 입력은 필수입니다.')
      }
      if (!this.model.end) {
        this.errors.push('종료날짜 입력은 필수입니다.')
      }
      return false
    },
    handleKeyDown (e) {
      const keyCode = e.keyCode
      if (keyCode === 27) {
        this.close()
      }
    },
    handleClick (e) {
      // FIXME: date-picker close
      if (!e.target.closest('.modal__inner')) {
        this.close()
      }
    },
    addEventListeners () {
      document.addEventListener('keydown', this.handleKeyDown)
      document.addEventListener('click', this.handleClick)
    },
    removeEventListeners () {
      document.addEventListener('keydown', this.handleKeyDown)
      document.addEventListener('click', this.handleClick)
    }
  }
}
</script>

<style lang="scss" scoped>
  .modal {
    opacity: 0;
    visibility: hidden;
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    text-align: left;
    background: rgba(0, 0, 0, .5);
    transition: opacity .25s ease;
  }

  .modal__bg {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    cursor: pointer;
  }

  .modal-state {
    display: none;
  }

  .show  {
    opacity: 1;
    visibility: visible;
  }

  .modal__inner {
    transition: top .25s ease;
    position: absolute;
    top: -20%;
    right: 0;
    bottom: 0;
    left: 0;
    width: 50%;
    margin: auto;
    overflow: auto;
    background: #fff;
    border-radius: 5px;
    padding: 1em 2em;
    height: 700px;

    & .form {
      margin-top: 1em;
      margin-bottom: 1em;
      input {
        width: 100%;
        padding: 12px 20px;
        margin: 8px 0;
        display: inline-block;
        border: 1px solid #ccc;
        border-radius: 4px;
        box-sizing: border-box;
        font-size: 14px;
      }
    }
    .validation {
      color: var(--color-accent);
      font-size: .8em;
    }
    .modal__action {
      display: flex;
      justify-content: flex-end;
      > button {
        border: none;
        background-color: var(--color-gray);
        padding: 16px 32px;
        text-decoration: none;
        margin: 4px 2px;
        cursor: pointer;
        &.info {
          background-color: var(--color-info);
          color: white;
        }
        &.accent {
          background-color: var(--color-accent);
          color: white;
        }
      }
    }
  }

  .modal__close {
    position: absolute;
    right: 1em;
    top: 1em;
    width: 1.1em;
    height: 1.1em;
    cursor: pointer;
  }

  .modal__close:after,
  .modal__close:before {
    content: '';
    position: absolute;
    width: 2px;
    height: 1.5em;
    background: #ccc;
    display: block;
    transform: rotate(45deg);
    left: 50%;
    margin: -3px 0 0 -1px;
    top: 0;
  }

  .modal__close:hover:after,
  .modal__close:hover:before {
    background: #aaa;
  }

  .modal__close:before {
    transform: rotate(-45deg);
  }
</style>
