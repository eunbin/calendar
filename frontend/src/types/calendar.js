const MONTH = 'month'
const WEEK = 'week'

const viewTypes = Object.freeze({
  MONTH,
  WEEK
})

export {
  viewTypes
}
