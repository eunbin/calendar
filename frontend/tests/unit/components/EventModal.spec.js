import { mount } from '@vue/test-utils'
import EventModal from '@/components/EventModal'
import dateMixin from '@/mixins/date'
import { modalTitles } from '@/types/calendar'
import { newEvent, eventHasId } from '../test-data'

const createOption = (event) => {
  return {
    mixins: [dateMixin],
    propsData: {
      value: true, // open modal,
      event
    }
  }
}

describe('EventModal.vue', () => {
  it('snapshot check with model', () => {
    const wrapper = mount(EventModal, createOption(newEvent))
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it('event model의 id 가 없는 경우는 일정 등록 타이틀을 나타낸다.', () => {
    const option = createOption(newEvent)

    const wrapper = mount(EventModal, option)// then

    expect(wrapper.vm.isAdd).toBeTruthy()
    expect(wrapper.vm.modalTitle).toMatch(modalTitles.ADD_EVENT)
  })

  it('event model의 id 가 있는 경우는 일정 수정 타이틀을 나타낸다.', () => {
    const option = createOption(eventHasId)

    const wrapper = mount(EventModal, option)

    expect(wrapper.vm.isAdd).toBe(false)
    expect(wrapper.vm.modalTitle).toMatch(modalTitles.UPDATE_EVENT)
  })

  it('일정 추가인 경우 수정 버튼은 렌더링되지 않는다.', () => {
    const option = createOption(newEvent)

    const wrapper = mount(EventModal, option)
    const saveButton = wrapper.find('.accent')

    expect(saveButton.exists()).toBe(false)
  })

  it('일정 수정인 경우 수정 버튼이 렌더링된다.', () => {
    const option = createOption(eventHasId)

    const wrapper = mount(EventModal, option)
    const saveButton = wrapper.find('.accent')

    expect(saveButton.exists()).toBe(true)
  })
})
