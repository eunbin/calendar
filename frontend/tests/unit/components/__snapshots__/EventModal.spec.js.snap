// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventModal.vue snapshot check with model 1`] = `
<div
  class="modal show"
  role="dialog"
>
  <div
    class="modal__inner"
  >
    <button
      class="modal__close"
    />
     
    <h2>
      일정 등록
    </h2>
     
    <div
      class="form"
    >
      <div>
        <label
          for="title"
        >
          일정 제목
        </label>
         
        <input
          id="title"
          maxlength="30"
          type="text"
        />
      </div>
       
      <div>
        <label
          for="startDate"
        >
          시작 시간
        </label>
         
        <div
          class="vdpComponent vdpWithInput"
          id="startDate"
        >
          <input
            type="text"
          />
           
          <button
            class="vdpClearInput"
            type="button"
          />
           
          <!---->
        </div>
      </div>
       
      <div>
        <label
          for="endDate"
        >
          종료 시간
        </label>
         
        <div
          class="vdpComponent vdpWithInput"
          id="endDate"
        >
          <input
            type="text"
          />
           
          <button
            class="vdpClearInput"
            type="button"
          />
           
          <!---->
        </div>
      </div>
    </div>
     
    <!---->
     
    <div
      class="modal__action"
    >
      <button>
        
        취소
      
      </button>
       
      <!---->
       
      <button
        class="info"
      >
        
        등록
      
      </button>
    </div>
  </div>
</div>
`;
